dimod/binary_quadratic_model.py

```diff
@@ -1612,6 +1612,92 @@ def from_json(cls, obj):
 
         return json.load(obj,  object_hook=bqm_decode_hook)
 
+    def to_bson(self, fp=None):
+        """Convert a binary quadratic model to an effecient BSON format.
+
+        Args:
+            fp (file, optional):
+                A `.write()`-supporting `file object`_ to save the binary 
+                quadratic model encoded in BSON format to. If not provided, 
+                returns bytes.
+        """
+                
+
+        import bson
+
+        num_variables = len(self)
+        if num_variables > 2**16:
+            raise ValueError
+
+        variable_order = sorted(self.linear)
+        num_possible_edges = num_variables*(num_variables - 1) // 2
+        density = len(self.quadratic) / float(num_possible_edges)
+        as_complete = density >= 0.5
+
+        lin, (i, j, _vals), off = self.to_numpy_vectors(
+            dtype=np.float32,
+            index_dtype=np.uint16,
+            sort_indices=as_complete,
+            variable_order=variable_order)
+
+        edge_idxs = i*(num_variables - 1) - i*(i+1)//2 + j - 1
+        if as_complete:
+            vals = np.zeros(num_possible_edges, dtype=np.float32)
+            vals[edge_idxs] = _vals
+
+        else:
+            vals = _vals
+
+        doc = {
+            "as_complete": as_complete,
+            "linear": bson.binary.Binary(lin.tobytes()),
+            "quadratic_vals": bson.binary.Binary(vals.tobytes()),
+            "vartype": "SPIN" if self.vartype == self.SPIN else "BINARY",
+            "offset": off,
+            "variable_order": variable_order,
+        }
+
+        if not as_complete:
+            doc["quadratic_head"] = bson.binary.Binary(i.tobytes())
+            doc["quadratic_tail"] = bson.binary.Binary(j.tobytes())
+
+        enc_doc = bson.BSON.encode(doc)
+        if fp is None:
+            return enc_doc
+
+        fp.write(enc_doc)
+
+    @classmethod
+    def from_bson(cls, obj):
+        """Deserialize a binary quadratic model from a BSON encoding.
+
+        Args:
+            obj: (bytes/file):
+                Either bytes or a  `.read()`-supporting `file object`_
+                that represents linear and quadratic biases using the encoding
+                used in `.to_bson`.
+        """
+
```

Comment:
Same argument as for `to_json` -- why accept files **and** bytes? Choose one. (Otherwise the game of `isinstance` can escalate quickly.) Preferably, be consistent with the rest of the API and the complementary method in particular.